File: monitors/barrier.py
#! /usr/bin/env python

import threading
import time
import random

THREADS =   10
PHASES  =   20

class Barrier(object):
    def __init__(self, n):
        self.arrived = 0
        self.n = n
        self.mutex = threading.Lock()
        self.allArrived = threading.Condition(self.mutex)

    def barrier(self):
        with self.mutex:
            self.arrived += 1

            if self.arrived == self.n:
                self.arrived = 0
                self.allArrived.notify_all()
            else:
                self.allArrived.wait()


def thread(barrier):
    id = threading.current_thread().name
    print("Start thread {}".format(id))

    for i in range(PHASES):
        time.sleep(random.random())
        barrier.barrier()
        print("{} finished phase {}".format(id, i))

    barrier.barrier()
    print("Finished thread {}".format(id))


def main():
    threads = []

    barrier = Barrier(THREADS)
    for i in range(THREADS):
        # Create new threads
        t = threading.Thread(target=thread, args=(barrier,))
        threads.append(t)

    # Start all threads
    for t in threads:
        t.start()

    # Wait for all threads to complete
    for t in threads:
        t.join()



if __name__ == "__main__":
    main()

File: monitors/Barrier.java
import java.util.Random;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class Barrier implements Runnable {
    static final int THREADS = 10;
    static final int PHASES  = 20;

    static volatile int arrived = 0;

    synchronized void barrier(int n) {
        arrived++;
        if (arrived == n) {
            arrived = 0;
            notifyAll();
        } else {
            try {
                wait();
            } catch (InterruptedException e) {}
        }
    }


    @Override
    public void run() {
        long id = Thread.currentThread().getId();
        Random rnd = new Random();

        System.out.printf("Thread %d\n", id);

        for (int i = 0; i < PHASES; i++) {
            try {
                Thread.sleep(rnd.nextInt(1000));
            } catch (InterruptedException e) {}

            barrier(THREADS);
            System.out.printf("%d finished phase %d\n", id, i);
        }
        barrier(THREADS);
        System.out.printf("Finished thread %d\n", id);

    }

    public static void main(String[] args) throws InterruptedException {
        Thread[] threads = new Thread[THREADS];
        int i;

        Barrier c = new Barrier();

        for (i=0; i< THREADS; i++) {
            threads[i] = new Thread(c);
            threads[i].start();
        }

        for (i=0; i< THREADS; i++) {
            threads[i].join();
        }
    }
}
